### PR TITLE
[TS] LPS-171662 Update logic updateOAuthClientEntry function

### PR DESCRIPTION
**Pull request by locpham97**
Hi team,
This PR is for ticket https://issues.liferay.com/browse/LPS-171662, which resolves [LPP-47525](https://issues.liferay.com/browse/LPP-47525). Please help me review it.
I have handled this ticket since the previous PR: https://github.com/liferay-appsec/liferay-portal/pull/960
Thanks,
Loc.
cc: @taiphamlr

### CODE REVIEW

File: modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java [hunk 1]
```diff
@@ -475,31 +506,37 @@ private void _updateOAuthClientEntry(
...
-						_generateAuthRequestParametersJSON(properties),
-						authServerWellKnownURI, _generateInfoJSON(properties),
-						OAuthClientEntryConstants.OIDC_USER_INFO_MAPPER_JSON,
-						_generateTokenRequestParametersJSON(properties));
+			OAuthClientEntry oAuthClientEntry = null;
+
+			if (oldProperties != null) {
+				String oldAuthServerWellKnownURI =
+					_updateOAuthClientASLocalMetadata(
+						defaultUserId, properties);
```
Comment (oliverkecskemety) [inline modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java:514]: For me it's a bit strange to call the variable `oldAuthServerWellKnownURI` but using `properties` to get it. I would assume to use `oldProperties` or not call it `old`.

File: modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java [hunk 2]
```diff
@@ -475,31 +506,37 @@ private void _updateOAuthClientEntry(
...
+							oldOAuthClientEntry.getOAuthClientEntryId(),
+							_generateAuthRequestParametersJSON(properties),
+							oldAuthServerWellKnownURI,
+							_generateInfoJSON(properties),
+							oldOAuthClientEntry.getOIDCUserInfoMapperJSON(),
+							_generateTokenRequestParametersJSON(properties));
+				}
+				else {
+					oAuthClientEntry = _addOrUpdateOAuthClientEntry(
+						companyId, properties, defaultUserId);
```
Comment (oliverkecskemety) [inline modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java:534]: Do you know if it's a valid case that we don't have `oldProperties` but we need to update an object? I'm not sure but if it's impossible than we can eliminate `_addOrUpdateOAuthClientEntry` and use only the add parts instead.

**Review by oliverkecskemety — COMMENTED** [review]
Hi @locpham97,

Thank you for the PR! Please check my comments!

**Comment by TaiPhamLR** [comment]
Thanks for your review @oliverkecskemety. I fixed it. I'm waiting for the next comments.

File: modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java [hunk 3]
```diff
@@ -475,31 +491,41 @@ private void _updateOAuthClientEntry(
...
+							oldOAuthClientEntry.getOAuthClientEntryId(),
+							_generateAuthRequestParametersJSON(properties),
+							authServerWellKnownURI,
+							_generateInfoJSON(properties),
+							oldOAuthClientEntry.getOIDCUserInfoMapperJSON(),
+							_generateTokenRequestParametersJSON(properties));
+				}
+				else {
+					oAuthClientEntry = _addOAuthClientEntry(
+						properties, defaultUserId);
```
Comment (TaiPhamLR) [inline modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java:523]: If we create a "Provider Connection" with `null` "OpenID Connect Client ID" and then update the entry with valid "OpenID Connect Client ID", the `oldProperties` is not `null` but `oldOAuthClientEntry` is null. A new `OAuthClientEntry` entry would be created.

File: modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java [hunk 4]
```diff
@@ -475,31 +489,38 @@ private void _updateOAuthClientEntry(
...
+							_generateAuthRequestParametersJSON(properties),
+							_updateOAuthClientASLocalMetadata(
+								defaultUserId, properties),
+							_generateInfoJSON(properties),
+							oldOAuthClientEntry.getOIDCUserInfoMapperJSON(),
+							_generateTokenRequestParametersJSON(properties));
+				}
+				else {
+					oAuthClientEntry = _addOAuthClientEntry(
+						properties, defaultUserId);
```
Comment (TaiPhamLR) [inline modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java:518]: If we create a "Provider Connection" with `null` "OpenID Connect Client ID" and then update the entry with valid "OpenID Connect Client ID", the `oldProperties` is not `null` but `oldOAuthClientEntry` is null. A new `OAuthClientEntry` entry would be created.

Comment (oliverkecskemety) [reply modules/apps/portal-security-sso/portal-security-sso-openid-connect-impl/src/main/java/com/liferay/portal/security/sso/openid/connect/internal/configuration/admin/service/OpenIdConnectProviderManagedServiceFactory.java:518]: Thanks for the clarification on this one. I checked it and I'll create a ticket for this one as it's not really intuitive for the user to understand why does it not work.

**Comment by oliverkecskemety** [comment]
ci:test:sf

**Comment by oliverkecskemety** [comment]
ci:test:relevant

**Review by oliverkecskemety — APPROVED** [review]
Hi @TaiPhamLR and @locpham97,

Thanks for the changes! Everything looks good to me now.

**Comment by oliverkecskemety** [comment]
ci:forward

**Pull request closed without merging**